compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FakeInliningLocalVariablesLowering.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.codegen.mapClass
+import org.jetbrains.kotlin.ir.IrElement
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrFunction
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.util.parentAsClass
+import org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid
+import org.jetbrains.kotlin.ir.visitors.acceptChildrenVoid
+import org.jetbrains.kotlin.load.java.JvmAbi
+
+internal val fakeInliningLocalVariablesLowering = makeIrFilePhase(
+    ::FakeInliningLocalVariablesLowering,
+    name = "FakeInliningLocalVariablesLowering",
+    description = "Add fake locals to identify the range of inlined functions and lambdas"
+)
+
+internal class FakeInliningLocalVariablesLowering(val context: JvmBackendContext) : IrElementVisitorVoid, FileLoweringPass {
+    override fun lower(irFile: IrFile) {
+        irFile.acceptChildrenVoid(this)
+    }
+
+    override fun visitElement(element: IrElement) {
+        element.acceptChildrenVoid(this)
+    }
+
+    override fun visitCall(expression: IrCall) {
+        expression.acceptChildrenVoid(this)
+        val callee = expression.symbol.owner
+        if (callee.isInline) {
```

Comment:
Could we reuse IrInlineReferenceLocator here (maybe be with expanding `inlineReferences = mutableSetOf<IrCallableReference>()` to inlineReferences = mutableSetOf<IrCallableReference, IrFunction (called function)>())

Reply:
We can (@pyos suggested that as well offline). I uploaded a version that does that. I'm adding methods for handling callable reference inline function arguments and for handling lamdba inline function arguments. In the fake variable lowering I do not record them in the collections as they will not be used. I could avoid doing that and just record the callee as well for lambda params and iterate them after the visit. I prefer to not build the collections when not needed, but what do you think?

Reply:
Your solution looks much better! Thanks for pr! Will merge it after remote run